### PR TITLE
Update behavior of base bond interface with NetworkManager

### LINKED ISSUE #5387: Set L3 method of unused NetworkManager managed interfaces
# Bug report
There is an oddity which occurs when cloud-init is configured to use NetworkManager (instead of sysconfig) and a bonded interface is configured with an additional tagged VLAN on the bonded interface.  This configuration causes an undesired behavior where the underlying untagged `bond0` interface is active from an L3 ipv4/ipv6 perspective and requests an address via DHCP while `bond0.123` is configured statically as outlined via an OpenStack network_data.json configuration.

## Steps to reproduce the problem
Create an OpenStack node with the following network_data.json:
```
{
  "links": [
    {
      "id": "ens1f0np0",
      "name": "ens1f0np0",
      "type": "phy",
      "ethernet_mac_address": "xx:xx:xx:xx:xx:00",
      "mtu": 9000
    },
    {
      "id": "ens1f1np1",
      "name": "ens1f1np1",
      "type": "phy",
      "ethernet_mac_address": "xx:xx:xx:xx:xx:01",
      "mtu": 9000
    },
    {
      "id": "bond0",
      "name": "bond0",
      "type": "bond",
      "bond_links": [
        "ens1f0np0", "ens1f1np1"
      ],
      "mtu": 9000,
      "ethernet_mac_address": "xx:xx:xx:xx:xx:00",
      "bond_mode": "802.3ad",
      "bond_xmit_hash_policy": "layer3+4",
      "bond_miimon": 100
    },
    {
      "id": "bond0.123",
      "name": "bond0.123",
      "type": "vlan",
      "vlan_link": "bond0",
      "vlan_id": 123,
      "vlan_mac_address": "xx:xx:xx:xx:xx:00"
    }
  ],
  "networks": [
    {
      "id": "publicnet-ipv4",
      "type": "ipv4",
      "link": "bond0.123",
      "ip_address": "x.x.x.x",
      "netmask": "255.255.255.0",
      "routes": [
        {
          "network": "0.0.0.0",
          "netmask": "0.0.0.0",
          "gateway": "x.x.x.1",
          "services": [
              {
              "type": "dns",
              "address": "1.1.1.1"
              },
              {
              "type": "dns",
              "address": "8.8.8.8"
              }
          ]
        }
      ],
      "network_id": "00000000-0000-0000-0000-000000000000"
    }
  ],
  "services": [
  ]
}
```

cloud-init will write the following interface files for the NetworkManager controlled `bond0` interfaces:
/etc/NetworkManager/system-connections/cloud-init-bond0.nmconnection
```
# Generated by cloud-init. Changes will be lost.

[connection]
id=cloud-init bond0
uuid=54317911-f840-516b-a10d-82cb4c1f075c
autoconnect-priority=120
type=bond
interface-name=bond0

[user]
org.freedesktop.NetworkManager.origin=cloud-init

[bond]
mode=802.3ad
miimon=100
xmit_hash_policy=layer3+4
```
/etc/NetworkManager/system-connections/cloud-init-bond0.123.nmconnection
```
# Generated by cloud-init. Changes will be lost.

[connection]
id=cloud-init bond0.123
uuid=d4411fab-2938-5f5a-adef-d9074937f11d
autoconnect-priority=120
type=vlan
interface-name=bond0.123

[user]
org.freedesktop.NetworkManager.origin=cloud-init

[vlan]
id=123
parent=54317911-f840-516b-a10d-82cb4c1f075c

[ipv4]
method=manual
may-fail=false
address1=x.x.x.x/24
route1=0.0.0.0/0,x.x.x.1
```

Because there are no explicit L3 ipv4/ipv6 sections in `bond0.nmconnection`, the untagged `bond0` interface defaults to `method=auto` and requests an IPv4 address via DHCP.  It just so happens that the physical network on which this server resides does respond to DHCP requests on the untagged VLAN so the server receives an address on the untagged VLAN which conflicts with the static configuration for the tagged VLAN specified via network_data.json.

Unfortunately, there is no provision to explicitly configure an interface as `type: disabled` in either the OpenStack network_data.json or cloud-init network data v1 schemas:
https://github.com/openstack/ironic/blob/8086167054193c94afef830ef9eccd820dd10a14/ironic/api/controllers/v1/network-data-schema.json#L272C1-L296C7
https://github.com/canonical/cloud-init/blob/e843f6a0c95f6137d4fa14f674e3126e09487abd/cloudinit/config/schemas/schema-network-config-v1.json#L479C1-L488C12

Perhaps there should be additional logic for certain types of interfaces to explicitly disable their L3 configuration?

## Environment details
- Cloud-init version: 23.4-7.el9_4.0.1
- Operating System Distribution: Rocky 9
- Cloud provider, platform or installer type: OpenStack Caracal 2024.1

### PR DESCRIPTION
**Pull request by jcmoore3**
When using NetworkManager, if the base bond interface does not have subnet information configured, ensure it is disabled with respect to ipv4 and ipv6. Otherwise, the base bond interface defaults to 'auto' and will try to configure itself via DHCP. This is problematic when using a tagged VLAN interface on top of the bond as the base interface will try to configure itself via DHCP on the untagged VLAN.

Fixes #5387 

## Proposed Commit Message

```
fix: Update behavior of base bond interface with NetworkManager

When using NetworkManager, if the base bond interface does not have
subnet information configured, ensure it is disabled with respect to
ipv4 and ipv6. Otherwise, the base bond interface defaults to 'auto'
and will try to configure itself via DHCP. This is problematic when
using a tagged VLAN interface on top of the bond as the base
interface will try to configure itself via DHCP on the untagged VLAN.

```

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [ ] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by jcmoore3** [comment]
Apologies that the CLA is taking so long. This needs to go through my company's internal open source approval process. I do not anticipate any issues, it just takes time. I'm hoping to have final internal approval this week.

**Comment by jcmoore3** [comment]
CLA signed, please let me know if anything else is required to get this moving. Change to github-cla-signers is in PR #5369.

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks, @jcmoore3, for fixing this. The changes look good to me, but unit tests need to be adapted and the new paths need to be covered.

**Comment by jcmoore3** [comment]
@aciba90 I have updated the unit tests but please note that there is a dependency on the change in #5496. Without this change, cloud-init current configures DNS settings for disabled interfaces which disallows NetworkManager from activating the interface.

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks for adding unit-tests! The problem you are describing happens if:

1. Base bond interface without ipvX.method defined defaulted to `auto`
2. VLAN linked interfaces with ipvX.method set to `manual`

Your PR proposes to change the default from `auto` to `disabled` for every interface without subnets. I have in-line requested to limit it, even more, to "bond" only interfaces.

In order to assess whether this fix should be backported or not, I am trying to think if cloud-init changes behavior with this change, as, for the described cases, we are fixing a bug, but for other cases, we are changing the default and possibly breaking other use-cases.

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks for limiting the change to bond interfaces. After that, the changed unittests are not relevant because they are xkipped, so could you please remove them applying the following patch:

<details>
  <summary>remove-tests.patch</summary>
 
  ```patch
commit 976707b7ca6d53b920cc2d82418f719b5e3f4da6
Author: Alberto Contreras <alberto.contreras@canonical.com>
Date:   Fri Jul 12 15:17:32 2024 +0200

    revert not needed tests changes
    
    line changes
    tests that are marked as xfail

diff --git a/tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection b/tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection
index 57222589f..fbdfbc652 100644
--- a/tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection
+++ b/tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection
@@ -11,4 +11,3 @@ interface-name=encc000
 org.freedesktop.NetworkManager.origin=cloud-init
 
 [ethernet]
-
diff --git a/tests/unittests/net/network_configs.py b/tests/unittests/net/network_configs.py
index 80c36dd0b..b68319cc8 100644
--- a/tests/unittests/net/network_configs.py
+++ b/tests/unittests/net/network_configs.py
@@ -1943,14 +1943,6 @@ pre-down route del -net 10.0.0.0/8 gw 11.0.0.1 metric 3 || true
                 [ethernet]
                 mac-address=66:BB:9F:2C:E8:80
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-eth5.nmconnection": textwrap.dedent(
@@ -2066,14 +2058,6 @@ pre-down route del -net 10.0.0.0/8 gw 11.0.0.1 metric 3 || true
                 [ethernet]
                 mac-address=98:BB:9F:2C:E8:80
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-eth1.nmconnection": textwrap.dedent(
@@ -2094,14 +2078,6 @@ pre-down route del -net 10.0.0.0/8 gw 11.0.0.1 metric 3 || true
                 [ethernet]
                 mac-address=AA:D6:9F:2C:E8:80
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-br0.nmconnection": textwrap.dedent(
@@ -2709,14 +2685,6 @@ pre-down route del -net 10.0.0.0/8 gw 11.0.0.1 metric 3 || true
                 [ethernet]
                 mac-address=66:BB:9F:2C:E8:80
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-eth5.nmconnection": textwrap.dedent(
@@ -3710,14 +3678,6 @@ iface bond0 inet6 static
                 [ethernet]
                 mac-address=AA:BB:CC:DD:E8:00
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-bond0s1.nmconnection": textwrap.dedent(
@@ -3738,14 +3698,6 @@ iface bond0 inet6 static
                 [ethernet]
                 mac-address=AA:BB:CC:DD:E8:01
 
-                [ipv4]
-                method=disabled
-                may-fail=false
-
-                [ipv6]
-                method=disabled
-                may-fail=false
-
                 """
             ),
             "cloud-init-bond0.nmconnection": textwrap.dedent(
  ```
</details>

As this PR depends on https://github.com/canonical/cloud-init/pull/5496, and that one will add unit tests, let's wait until 5496 is merged and add the tests after a rebase.

**Comment by jcmoore3** [comment]
@aciba90 The tests have been updated as suggested

File: tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection
Comment (aciba90) [inline tests/unittests/net/artifacts/no_matching_mac/etc/NetworkManager/system-connections/cloud-init-encc000.nmconnection:14]: This extra line is unneeded, could please remove it?

**Review by aciba90 — CHANGES_REQUESTED** [review]
Could you please rebase main to include https://github.com/canonical/cloud-init/pull/5496 and include unit tests covering this change ?

**Review by aciba90 — CHANGES_REQUESTED** [review]
Thanks for rebasing `main` and adapting the existent tests. Could you please include unit tests covering this change?

**Review by aciba90 — APPROVED** [review]
LGTM, I see now that the unit tests cover the change, no need to add extra ones. Thanks!

**Pull request merged**
